eureka/S3_data_reduction/wfc3.py

```diff
@@ -0,0 +1,498 @@
+
+    Returns
+    -------
+    data: DataClass
+        The updated data object with the fits data stored inside
+
+    Notes
+    -----
+    History:
+
+    - January 2017 Kevin Stevenson
+            Initial code as implemented in the WFC3 pipeline
+    - 18-19 Nov 2021 Taylor Bell
+        Edited and decomposed WFC3 code to integrate with Eureka!          
+    '''
+    assert isinstance(filename, str)
+
+    #Determine image size and filter/grism
+    with fits.open(filename) as hdulist:
+        data.filename   = filename
+        data.mhdr       = hdulist[0].header
+        data.shdr       = hdulist[1].header
+        meta.nx         = data.shdr['NAXIS1']
+        meta.ny         = data.shdr['NAXIS2']
+        meta.grism      = data.mhdr['FILTER']
+        meta.detector   = data.mhdr['DETECTOR']
+        meta.flatoffset = [[-1*data.shdr['LTV2'], -1*data.shdr['LTV1']]]
+        
+        # Determine if we are using IMA or FLT files
+        if filename.endswith('flt.fits'):
+            # FLT files subtract first from last, 2 reads
+            meta.nreads = 2
+        else:
+            meta.nreads = data.shdr['SAMPNUM']
+
+        if data.shdr['BUNIT'] == 'ELECTRONS/S':
+            #Science data and uncertainties were previously in units of e-/sec,
+            #therefore multiply by sample time to get electrons.
+            samptime = data.shdr['SAMPTIME']
+        else:
+            samptime = 1
+
+        data.data = np.zeros((meta.nreads,meta.ny,meta.nx))    #Flux
+        data.err = np.zeros((meta.nreads,meta.ny,meta.nx))    #Flux
+        data.dq = np.zeros((meta.nreads,meta.ny,meta.nx))    #Flux
+        jd = []
+        j = 0
+        for rd in range(meta.nreads,0,-1):
+            data.data[j] = hdulist['SCI',rd].data*samptime
+            data.err[j]  = hdulist['ERR',rd].data*samptime
+            data.dq[j]   = hdulist['DQ', rd].data
+            jd.append(2400000.5 + hdulist['SCI',rd].header['ROUTTIME']-0.5*hdulist['SCI',rd].header['DELTATIM']/3600/24)
+            j += 1
+        jd = np.array(jd)
+
+        # data.data = hdulist[1].data*samptime
+        # data.err  = hdulist[2].data*samptime
+        # data.dq  = hdulist[3].data
+    
+    # jd  = np.array([2400000.5 + 0.5*(data.mhdr['EXPSTART'] + data.mhdr['EXPEND'])])
+    ra  = data.mhdr['RA_TARG']*np.pi/180
+    dec = data.mhdr['DEC_TARG']*np.pi/180
+    if meta.horizonsfile != None:
+        # Apply light-time correction, convert to BJD_TDB
+        # Horizons file created for HST around time of observations
+        bjd_corr = suntimecorr.suntimecorr(ra, dec, jd, meta.horizonsfile)
+        bjdutc      = jd + bjd_corr/86400.
+        data.bjdtdb = bjdutc # FINDME: Update to the lines below after I have the leapdir files
+        #data.bjdtdb   = utc_tt.utc_tt(bjdutc,meta.leapdir)   # FIX: should this be utc_tdb instead?
+    elif meta.firstFile:
+        print("WARNING: No Horizons file found. Using JD rather than BJD_TDB.")
+        data.bjdtdb   = jd
```

Comment:
Having a variable called `bjdtdb` but it actually being `jd` is a recipe for disaster. Maybe we use `hjdutc` or simply `jd` and leave `bjdtdb` undefined? This may cause headaches later on in the code

Reply:
The issue with that is that we'd need an if/else statement everywhere afterward that uses jd if bjdtdb is undefined. A better option might be to re-write the code for all the instruments to simply use a `time` variable that has an associated `time_units` variable which can be checked to see if it is bdjtdb or jd or whatever. Thoughts?

Reply:
That's how POET operates, where the time unit can also be orbital phase. Let's do it!

Reply:
Working on this right now - I'll push some commits soon and then we can merge this branch into main. Just testing end-to-end with NIRSpec and WFC3 to make sure I haven't broken anything with this time and time_units change